src/Plugin/EntityBrowser/Widget/Webdam.php

```diff
@@ -0,0 +1,286 @@
+        //Add the current folder to the breadcrumb
+        $breadcrumbs[$current_folder_id] = $current_folder->name;
+      }
+    }
+    //Reset the breadcrumb array so that it can be rendered in order
+    reset($breadcrumbs);
+    //Create a container for the breadcrumb
+    $form['breadcrumb-container'] = [
+      '#type' => 'container',
+      //custom element property to store breadcrumbs array.  This is fetched from the form state every time the form is rebuilt due to navigating between folders
+      '#breadcrumbs' => $breadcrumbs,
+    ];
+    //Add the breadcrumb buttons to the form
+    foreach ($breadcrumbs as $folder_id => $folder_name){
+      $form['breadcrumb-container'][$folder_id] = [
+        '#type' => 'button',
+        '#value' => $folder_name,
+        '#name' => 'webdam_folder',
+        '#webdam_folder_id' => $folder_id,
+        '#webdam_parent_folder_id' => $folder_name,
+      ];
+    }
+
+    //Add container for assets (and folder buttons)
+    $form['asset-container'] = [
+      '#type' => 'container',
+    ];
+    //Add folder buttons to form
+    foreach ($folders as $folder){
+      $form['asset-container'][$folder->id] = [
+        '#type' => 'button',
+        '#value' => $folder->name,
+        '#name' => 'webdam_folder',
+        '#webdam_folder_id' => $folder->id,
+        '#webdam_parent_folder_id' => $current_folder->parent,
+      ];
+    }
+    //Assets are rendered as #options for a checkboxes element.  Start with an empty array.
+    $assets = [];
+    //Add to the assets array
+    foreach ($folder_items as $folder_item) {
+      if($folder_item->type == 'asset'){
+        //wrap asset name in a paragraph tag
+        $assets[$folder_item->id] = '<p class="webdam-asset-fielname">'.$folder_item->name.'</p>';
+        //If the asset has a thumbnail it should be rendered
+        if(!empty($folder_item->thumbnailurls)){
+          //Multiple thumbnail sizes are available.  Using the 220 width version for now
+          $assets[$folder_item->id] .= '<img class="webdam-asset-thumbnail" src="'.$folder_item->thumbnailurls[2]->url.'" alt="'.$folder_item->name.'" />';
+        }
+      }
+    }
+    //If the assets array is not empty then add the assets to the form as checkboxes
+    if(!empty($assets)){
+      $form['asset-container']['assets'] = [
+        '#type' => 'checkboxes',
+        '#title' => $this->t('Choose one or more assets'),
+        '#title_display' => 'invisible',
+        '#options' => $assets,
+        // Multiple assets will only be accepted if the source field allows more than one value.
+        '#multiple' => $field_cardinality != 1 && $this->configuration['multiple'],
+      ];
+    }
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function prepareEntities(array $form, FormStateInterface $form_state) {
+    foreach ($form_state->getValue(['upload'], []) as $aid) {
+      if ($aid !== 0) {
+        //TODO: validate form selection
```

Comment:
This seems somewhat important. Should we do this now?

Reply:
The only validation I can think of is to fetch all assets from webdam to verify that they are valid assets.  In order to do that I think we would need to implement https://www.damsuccess.com/hc/en-us/articles/202134055-REST-API#getmultiasset into the webdam client.  What do you think?  Is there other validation you can think of?

Reply:
Ah, I see what you're saying. Yeah, skip it for now. We can circle back to it.

Reply:
Opened an issue in github to track it.